spiel/cmd.c

```diff
@@ -2025,30 +2044,195 @@ static int spiel__proc_is_to_update_stats(struct spiel_proc_item *proc,
+	struct m0_fop                  *fop = &proc->sci_fop;
+	struct m0_rpc_item             *item;
+	m0_time_t                       conn_timeout;
+	int                             rc;
+
+	M0_ENTRY(FID_F, FID_P(&proc->sci_fid));
+	m0_clink_fini(clink);
+	if (proc->sci_rlink.rlk_rc != 0) {
+		M0_LOG(M0_ERROR, "connect failed");
+		M0_LEAVE();
+		return true;
+	}
+	m0_fop_init(fop, &m0_fop_process_fopt, NULL, m0_fop_release);
+	rc = m0_fop_data_alloc(fop);
+	if (rc != 0) {
+		M0_LOG(M0_ERROR, "fop data alloc failed");
+		goto fop_fini;
+	}
+
+	fop->f_item.ri_rmachine = proc->sci_spc->spc_rmachine;
+	req                     = m0_ss_fop_process_req(fop);
+	req->ssp_cmd            = M0_PROCESS_COUNTER;
+	req->ssp_id             = proc->sci_fid;
+	item                    = &fop->f_item;
+	item->ri_ops            = &spiel_process_counter_ops;
+	item->ri_session        = &proc->sci_rlink.rlk_sess;
+	item->ri_prio           = M0_RPC_ITEM_PRIO_MID;
+	item->ri_nr_sent_max    = 5;
+	m0_fop_get(fop);
+	rc = m0_rpc_post(item);
+	
+	if (rc != 0) {
+		M0_LOG(M0_ERROR, "rpc post failed");
+		goto fop_put;
+	}
+
+	M0_LEAVE();
+	return true;
+fop_put:
+	m0_fop_put_lock(fop);
+fop_fini:
+	m0_fop_fini(fop);
+	conn_timeout = m0_time_from_now(SPIEL_CONN_TIMEOUT, 0);
+	m0_rpc_link_disconnect_sync(&proc->sci_rlink, conn_timeout);
+	M0_LEAVE();
+	return true;
+}
+
+static int spiel_process__counters_async(struct spiel_proc_counter_item *proc)
+{
+	struct m0_conf_process *process;
+	struct m0_spiel_core   *spc = proc->sci_spc;
+	m0_time_t               conn_timeout;
+	int                     rc;
+
+	M0_ENTRY("proc fid "FID_F, FID_P(&proc->sci_fid));
+	rc = spiel_proc_conf_obj_find(spc, &proc->sci_fid, &process);
+	if (rc != 0)
+		return M0_ERR(rc);
+	m0_confc_close(&process->pc_obj);
+	conn_timeout = m0_time_from_now(SPIEL_CONN_TIMEOUT, 0);
+	rc = m0_rpc_link_init(&proc->sci_rlink, spc->spc_rmachine, NULL,
+			      process->pc_endpoint, SPIEL_MAX_RPCS_IN_FLIGHT);
+	if (rc != 0)
+		return M0_ERR(rc);
+
+	m0_clink_init(&proc->sci_rlink_wait, spiel_proc_counter_item_rlink_cb);
+	proc->sci_rlink_wait.cl_is_oneshot = true;
+	m0_rpc_link_connect_async(&proc->sci_rlink, conn_timeout,
+				  &proc->sci_rlink_wait);
+	return M0_RC(rc);
+}
+
 int m0_spiel_proc_counters_fetch(struct m0_spiel        *spl,
 				 struct m0_fid          *proc_fid,
 				 struct m0_proc_counter *count_stats)
 {
+	struct spiel_proc_counter_item *proc = NULL;
+	struct m0_spiel_core           *spc = &spl->spl_core;
+	struct m0_conf_obj             *proc_obj;
 	int                             rc = 0;
 
 	M0_ENTRY();
 	M0_PRE(spl != NULL);
+	M0_PRE(spc->spc_confc != NULL);
 	M0_PRE(count_stats != NULL);
 
+	if (!m0_confc_is_inited(spc->spc_confc))
+		return M0_ERR_INFO(-EAGAIN, "confc is finalised");
+
+	rc = m0_confc_open_by_fid_sync(spc->spc_confc, proc_fid, &proc_obj);
+	if (rc != 0)
+		return M0_ERR(rc);
+	if (proc_obj->co_ha_state != M0_NC_ONLINE) {
+		rc = M0_ERR(-EINVAL);
```

Comment:
For this error hare needs to use snapshot of old value.
Need to define proper error code for it.

Reply:
Done.